include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -861,111 +861,42 @@ class EnablePreconditionedIterativeSolver
 };
 
 
-/**
- * The parameter type shared between all iterative solvers.
- * @see GKO_CREATE_FACTORY_PARAMETERS
- */
-struct iterative_solver_factory_parameters {
+template <typename Parameters, typename Factory>
+struct enable_iterative_solver_factory_parameters
```

Comment:
Yes, with this change there is no single type that represents an interative solver's parameters, because all `enable_...` types are distinct due to their template parameters.

Reply:
```
void parse_iterative_solver_config(iterative_solver_factory_parameters& params, const config& c);
void dump_iterative_solver_config(const iterative_solver_factory_parameters& params, config& c);
```
No need to use templates in this case

Reply:
you can not do it just in this way, because there's no concrete type out of them.
we still need to know the what's the actual type, so it will still require the template.
I do not think two/three parameter affect the compile time a lot and the part is not in public interface

Reply:
yes you can :) https://godbolt.org/z/3Gb8TTbbM

Reply:
I know you can pass that like this way. like I said, it requires you know the actual type but it is done by the previous template from scratch.

Reply:
The following does not work.
```
iterative_parameter a;
parse(a, config);
a->generate() -> what type is it
```

Reply:
That's not the purpose of the type. It is meant to be usable by us and by users to generically access or set parameters for iterative solvers without knowing the actual type and without having to use templates.

In general, this PR is already going beyond what it aims to do (prevent future interface breaks for `deferred_factory_parameter`).

Reply:
IMO, the previous approach lead to a rather complicated hierarchy with very little gain from that. We don't currently need the common type of the parameters, and I don't expect that we will in the near future. Mike's approach is a lot cleaner, since the hierarchy is a single chain, while in the other case it's a tree.
Regarding the compile times, as long as a user compiles GPU kernels, this will have barely any impact.